Fenrir/Commands.fs

```diff
@@ -159,3 +165,86 @@ let stringToByte (s: String): byte[] =
             Byte.Parse(s.AsSpan(charIndex, 2), NumberStyles.AllowHexSpecifier, provider = CultureInfo.InvariantCulture)
         )
     | n -> failwithf "String of invalid length %d: %s" n s
+
+let hashOfObjectInTree (tree: TreeBody) (name: String): byte[] =
+    let atom = Array.find (fun a -> a.Name = name) tree
+    atom.Hash
+
+let changeHashInTree (tree: TreeBody) (hash: byte[]) (name: String): TreeBody =
+    let changer (a: TreeAtom) : TreeAtom =
+        match (a.Name = name) with
+            | true -> {Mode = a.Mode; Name = a.Name; Hash = hash}
+            | false -> a
+    Array.map changer tree
+
+let treeBodyToStream (tree: TreeBody) (stream: Stream): unit =
+    let printAtom (a : TreeAtom): unit =
+        stream.Write(ReadOnlySpan<byte>(a.Mode.ToString(CultureInfo.InvariantCulture) |> Encoding.ASCII.GetBytes))
+        stream.WriteByte(' 'B)
+        stream.Write(ReadOnlySpan<byte>(a.Name |> Encoding.ASCII.GetBytes))
+        stream.WriteByte(00uy)
+        stream.Write(ReadOnlySpan<byte>(a.Hash))
+    Array.iter printAtom tree
+
+let changeHashInCommit (commit: CommitBody) (hash: byte[]): CommitBody =
+    {Tree = hash |> byteToString;
+     Parents = commit.Parents;
+     Rest = commit.Rest}
+
+let commitBodyToStream (commit: CommitBody) (stream: Stream): unit =
+    let printParent (a: String): unit =
+        stream.Write(ReadOnlySpan<byte>("parent "B))
+        stream.Write(ReadOnlySpan<byte>(a |> Encoding.ASCII.GetBytes))
+        stream.WriteByte('\n'B)
+
+    stream.Write(ReadOnlySpan<byte>("tree "B))
+    stream.Write(ReadOnlySpan<byte>(commit.Tree |> Encoding.ASCII.GetBytes))
+    stream.WriteByte('\n'B)
+
+    Array.iter printParent commit.Parents
+    stream.Write(ReadOnlySpan<byte>(String.Join('\n', commit.Rest) |> Encoding.ASCII.GetBytes))
+
+type TreeStreams (length: int) =
+    member this.Streams = Array.create (length) (new MemoryStream())
+    member this.Hashes = Array.create (length) ""
+    interface System.IDisposable with
+        member this.Dispose() =
+            Array.iter (fun (s: MemoryStream) -> s.Dispose()) this.Streams
+
+let updateObjectInTree (rootTreeHash: string) (pathToRepo: String) (filePath: string) (blobHash: string): TreeStreams =
+    let filePathList = filePath.Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries) |> List.ofArray
+    use treeStreams = new TreeStreams (filePathList.Length)
+
+    let rec updateFileHashInTree (tree: TreeBody) (filePaths: String list): String =
+        let index = treeStreams.Streams.Length - filePaths.Length
+        match filePaths with
+        | [] -> failwithf "Empty path to file"
+        | [fileName] ->
+            let newTree = changeHashInTree tree (stringToByte blobHash) fileName
+
+            use input = new MemoryStream()
+            treeBodyToStream newTree input
+            writeObjectHeader GitObjectType.GitTree input treeStreams.Streams.[index]
```

Comment:
Could you help me please? This line isn't work, it doesn't write anything to MemoryStream. Did I code `type TreeStreams` properly?

Reply:
I partially fixed this problem. Now it looks like all members of MemoryStream[] points to one MemoryStream. I can send commit, if it has some sense.

I guess problem can be solved by drastic change of functions' types, namely by replacing the vast majority of Stream functions to byte[] functions. Only packing/unpacking functions require Stream objects, while others can be rewritten using byte[].

Reply:
Please see my review comments. I don't think there's immediate reason to replace the current abstraction.